samples/kotlin-android-app-destinations/src/main/java/com/segment/analytics/destinations/plugins/FirebaseDestination.kt

```diff
@@ -0,0 +1,236 @@
+                "Checkout Started" to Event.BEGIN_CHECKOUT,
+                "Order Completed" to Event.ECOMMERCE_PURCHASE,
+                "Order Refunded" to Event.PURCHASE_REFUND,
+                "Product Viewed" to Event.VIEW_ITEM,
+                "Product List Viewed" to Event.VIEW_ITEM_LIST,
+                "Payment Info Entered" to Event.ADD_PAYMENT_INFO,
+                "Promotion Viewed" to Event.PRESENT_OFFER,
+                "Product Added to Wishlist" to Event.ADD_TO_WISHLIST,
+                "Product Shared" to Event.SHARE,
+                "Product Clicked" to Event.SELECT_CONTENT,
+                "Products Searched" to Event.SEARCH
+        )
+
+        private val PROPERTY_MAPPER: Map<String, String> = mapOf(
+                "category" to Param.ITEM_CATEGORY,
+                "product_id" to Param.ITEM_ID,
+                "name" to Param.ITEM_NAME,
+                "price" to Param.PRICE,
+                "quantity" to Param.QUANTITY,
+                "query" to Param.SEARCH_TERM,
+                "shipping" to Param.SHIPPING,
+                "tax" to Param.TAX,
+                "total" to Param.VALUE,
+                "revenue" to Param.VALUE,
+                "order_id" to Param.TRANSACTION_ID,
+                "currency" to Param.CURRENCY,
+                "products" to Param.ITEM_LIST
+        )
+
+        private val PRODUCT_MAPPER: Map<String, String> = mapOf(
+                "category" to Param.ITEM_CATEGORY,
+                "product_id" to Param.ITEM_ID,
+                "id" to Param.ITEM_ID,
+                "name" to Param.ITEM_NAME,
+                "price" to Param.PRICE,
+                "quantity" to Param.QUANTITY
+        )
+    }
+
+    override fun setup(analytics: Analytics) {
+        super.setup(analytics)
+
+        firebaseAnalytics = FirebaseAnalytics.getInstance(context)
+    }
+
+    override fun identify(payload: IdentifyEvent): BaseEvent? {
+        var returnPayload = super.identify(payload)
+
+        firebaseAnalytics.setUserId(payload.userId)
+
+        payload.traits.let {
+            for ((traitKey, traitValue) in it) {
+                val updatedTrait = makeKey(traitValue.toString())
+                firebaseAnalytics.setUserProperty(traitKey, updatedTrait)
+            }
+        }
+
+        return returnPayload
+    }
+
+    override fun track(payload: TrackEvent): BaseEvent? {
+        var returnPayload = super.track(payload)
+
+        // Clean the eventName up
+        var eventName = payload.event
+        if (EVENT_MAPPER.containsKey(eventName)) {
+            eventName = EVENT_MAPPER[eventName].toString()
+        } else {
+            eventName = makeKey(eventName)
+        }
+
+        payload.properties.let {
+            val formattedProperties = formatProperties(it)
+            firebaseAnalytics.logEvent(eventName, formattedProperties)
+            analytics.log("firebaseAnalytics.logEvent($eventName, $formattedProperties")
```

Comment:
missing a trailing `)`

Reply:
Hah I was confused why it was compiling, but realized it was just in the string. Added.

Reply:
whoops, should have been clearer 😅